api/auth/external_auth.proto

```diff
@@ -0,0 +1,147 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2.auth;
+
+import "api/address.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with `OK` or `NOT OK`.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The request attributes.
+  AttributeContext attributes = 1;
+}
+
+message CheckResponse {
+  // Status `OK` allows the request. Any other status indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// AttributeContext is a collection of individual attributes.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The address of the peer, this is typically the IP address.
+    // It can also be UDS, or others.
+    Address address = 1;
+
+    // The canonical service name of the peer.
+    // It should be set to :ref:`x-envoy-downstream-service-cluster
+    // <https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/headers#x-envoy-downstream-service-cluster>`
+    // If a more trusted source of the service name is available through mTLS/secure naming, it
+    // should be used.
+    string service = 2;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
+    map<string, string> labels = 3;
+
+    // The authenticated identity of this peer.
+    // For example, the identity associated with the workload such as a service account.
+    // If an X.509 certificate is used to assert the identity this field should be sourced from
+    // `Subject` or `Subject Alternative Names`. The primary identity should be the principal.
+    // The principal format is issuer specific.
+    // For example
+    // SPIFFE format is `spiffe://trust-domain/path`
+    // Google account format is `https://accounts.google.com/{userid}`
+    string principal = 4;
+  }
+
+  // Represents a network request, such as an HTTP request.
+  message Request {
+    oneof request {
+      HTTPRequest http_request = 1;
```

Comment:
This has the same issue as above Address proto. Now the attribute name becomes request.http_request.path, which conflicts with Istio design. We could just rename Request to HTTPRequest and force every protocol to normalize its request to HTTP request.

We need to be aware, the attributes are designed for users to write policies against. We don't really care about how machine process it. We should note the number of HTTP requests far outweigh other requests.

Reply:
I think we have two alternatives.
1. Introduce canonical aliases as proto options.
2. Revert to original Request. Precisely specify mapping for non http or higher level protocols into the abstract Request message instead of leaving it to runtime.

Reply:
Sorry but I'm strongly against going back to what we have before. What we have here is more factual and easy for management servers to deal with. It's a set of facts. The management server can do whatever it wants.

@wora @mandarjog we need to move this forward to unblock the community. I'm sorry if this is not what you feel is optimal for Istio/Mixer, but I think it's optimal for the larger community. Thanks.